File: radvel/kepler.py

import numpy as np
import radvel

# Try to import Kepler's equation solver written in C
try:
    from . import _kepler 
    cext = True
except ImportError:
    print("WARNING: KEPLER: Unable to import C-based Kepler's\
equation solver. Falling back to the slower NumPy implementation.")
    cext = False


def rv_drive(t, orbel, use_c_kepler_solver=cext):
    """RV Drive
    
    Args:
        t (array of floats): times of observations
        orbel (array of floats): [per, tp, e, om, K].\
            Omega is expected to be\
            in radians
        use_c_kepler_solver (bool): (default: True) If \
            True use the Kepler solver written in C, else \
            use the Python/NumPy version.
    Returns:
        rv: (array of floats): radial velocity model
    
    """
    
    # unpack array of parameters
    per, tp, e, om, k = orbel
    
    # Performance boost for circular orbits
    if e == 0.0:
        m = 2 * np.pi * (((t - tp) / per) - np.floor((t - tp) / per))
        return k * np.cos(m + om)
    
    if per < 0:
        per = 1e-4
    if e < 0:
        e = 0
    if e > 0.99:
        e = 0.99

    # Calculate the approximate eccentric anomaly, E1, via the mean anomaly  M.
    if use_c_kepler_solver:
        rv = _kepler.rv_drive_array(t, per, tp, e, om, k)
    else:
        nu = radvel.orbit.true_anomaly(t, tp, per, e)
        rv = k * (np.cos(nu + om) + e * np.cos(om))
    
    return rv


def kepler(inbigM, inecc):
    """Solve Kepler's Equation

    Args:
        inbigM (array): input Mean anomaly
        inecc (array): eccentricity

    Returns:
        array: eccentric anomaly
    
    """
    
    Marr = inbigM  # protect inputs; necessary?
    eccarr = inecc
    conv = 1.0e-12  # convergence criterion
    k = 0.85

    Earr = Marr + np.sign(np.sin(Marr)) * k * eccarr  # first guess at E
    # fiarr should go to zero when converges
    fiarr = ( Earr - eccarr * np.sin(Earr) - Marr)  
    convd = np.where(np.abs(fiarr) > conv)[0]  # which indices have not converged
    nd = len(convd)  # number of unconverged elements
    count = 0

    while nd > 0:  # while unconverged elements exist
        count += 1
        
        M = Marr[convd]  # just the unconverged elements ...
        ecc = eccarr[convd]
        E = Earr[convd]

        fi = fiarr[convd]  # fi = E - e*np.sin(E)-M    ; should go to 0
        fip = 1 - ecc * np.cos(E)  # d/dE(fi) ;i.e.,  fi^(prime)
        fipp = ecc * np.sin(E)  # d/dE(d/dE(fi)) ;i.e.,  fi^(\prime\prime)
        fippp = 1 - fip  # d/dE(d/dE(d/dE(fi))) ;i.e.,  fi^(\prime\prime\prime)

        # first, second, and third order corrections to E
        d1 = -fi / fip 
        d2 = -fi / (fip + d1 * fipp / 2.0)
        d3 = -fi / (fip + d2 * fipp / 2.0 + d2 * d2 * fippp / 6.0)
        E = E + d3
        Earr[convd] = E
        fiarr = ( Earr - eccarr * np.sin( Earr ) - Marr) # how well did we do?
        convd = np.abs(fiarr) > conv  # test for convergence
        nd = np.sum(convd is True)
        
    if Earr.size > 1: 
        return Earr
    else: 
        return Earr[0]


def profile():
    # Profile and compare C-based Kepler solver with
    # Python/Numpy implementation

    import timeit
    
    ecc = 0.1
    numloops = 5000
    print("\nECCENTRICITY = {}".format(ecc))

    for size in [10, 30, 100, 300, 1000]:

        setup = """\
from radvel.kepler import rv_drive
import numpy as np
gc.enable()
ecc = %f
orbel = [32.468, 2456000, ecc, np.pi/2, 10.0]
t = np.linspace(2455000, 2457000, %d)
""" % (ecc, size)

        print("\nProfiling pure C code for an RV time series with {} "
              "observations".format(size))
        tc = timeit.timeit('rv_drive(t, orbel, use_c_kepler_solver=True)',
                           setup=setup, number=numloops)
        print("Ran %d model calculations in %5.3f seconds" % (numloops, tc))

        print("Profiling Python code for an RV time series with {} "
              "observations".format(size))
        tp = timeit.timeit('rv_drive(t, orbel, use_c_kepler_solver=False)',
                           setup=setup, number=numloops)
        print("Ran %d model calculations in %5.3f seconds" % (numloops, tp))
        print("The C version runs %5.2f times faster" % (tp/tc))

    ecc = 0.7
    numloops = 5000
    print("\nECCENTRICITY = {}".format(ecc))

    for size in [30]:
            setup = """\
from radvel.kepler import rv_drive
import numpy as np
gc.enable()
ecc = %f
orbel = [32.468, 2456000, ecc, np.pi/2, 10.0]
t = np.linspace(2455000, 2457000, %d)
    """ % (ecc, size)

            print("\nProfiling pure C code for an RV time series with {} "
                  "observations".format(size))
            tc = timeit.timeit('rv_drive(t, orbel, use_c_kepler_solver=True)',
                               setup=setup, number=numloops)
            print("Ran %d model calculations in %5.3f seconds" % (numloops, tc))

            print("Profiling Python code for an RV time series with {} "
                  "observations".format(size))
            tp = timeit.timeit('rv_drive(t, orbel, use_c_kepler_solver=False)',
                               setup=setup, number=numloops)
            print("Ran %d model calculations in %5.3f seconds" % (numloops, tp))
            print("The C version runs %5.2f times faster" % (tp / tc))

File: src/kepler.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

int sign(int x) {
  return (x > 0) - (x < 0);
}

/*
Solution to Kepler's equation. Given mean anomaly, M, and eccentricity, e,
solve for E, the eccentric anomaly, which must satisfy:

    E - e sin(E) - M = 0

Follows the method of Danby 1988 as written in Murray and Dermot p36-37.
*/

double kepler(double M, double e);
double kepler(double M, double e)
{
  int MAX_ITER = 30;
  double CONV_TOL = 1.0e-12; // convergence criterion

  double k, E, fi, d1, fip, fipp, fippp;
  int count;
  k = 0.85; // initial guess at input parameter
  count = 0; // how many loops have we done?

  E = M + sign(sin(M)) * k * e; // first guess at E, the eccentric anomaly

  // E - e * sin(E) - M should go to 0
  fi = (E - e * sin(E) - M); 
  while ( fabs(fi) > CONV_TOL && count < MAX_ITER)
    {
      count++;
      
      // first, second, and third order derivatives of fi with respect to E
      fip = 1 - e * cos(E) ;
      fipp = e * sin(E); 
      fippp = 1 - fip; 

      // first, second, and third order corrections to E
      d1 = -fi / fip; 
      d1 = -fi / (fip + d1 * fipp / 2.0); 
      d1 = -fi / (fip + d1 * fipp / 2.0 + d1 * d1 * fippp / 6.0);
      E += d1;

      fi  = (E - e * sin(E) - M);
      // printf("E =  %f, count = %i\n", E , count); //debugging

      if(count==MAX_ITER){
	printf("Error: kepler step not converging after %d steps.\n", MAX_ITER);
	printf("E=%f,  M=%f,  e=%f\n", E, M, e);
	exit(-1);
      }
    }
  return E;
}

// om is in radians
double rv_drive(double t, double per, double tp, double e, double cosom, double sinom, double k)
{
  double PI = 3.141592653589793;
  double phase, M, E, ratio,fac, rv;
  phase = (t - tp) / per;
  M = 2 * PI * ( phase - floor( phase ) );

  // Calculate the approximate eccentric anomaly, E1, via the mean anomaly  M.
  E = kepler(M, e);
    
  // Calculate the radial velocity
  ratio = sqrt((1 + e)/(1 - e))*tan ( E / 2 );
  fac = 2/(1 + ratio*ratio);
  rv = k * (cosom*(fac - 1) - ratio * fac * sinom + e*cosom);
  //printf("hello");
  return rv;
}

// little test function 
int main()
{
  double M, e, E;
  printf("Enter M, the mean anomally\n");
  scanf("%lf", &M);
  printf("Enter e, the eccentricity\n");
  scanf("%lf", &e);
  E = kepler(M, e);
  printf("%lf\n",E);
  return 0;
}
